src/python/pants/backend/cc/lint/clangformat/subsystem.py

```diff
@@ -0,0 +1,79 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import os
+from typing import Iterable
+
+from pants.backend.python.goals import lockfile
+from pants.backend.python.goals.lockfile import GeneratePythonLockfile
+from pants.backend.python.subsystems.python_tool_base import PythonToolBase
+from pants.backend.python.subsystems.setup import PythonSetup
+from pants.backend.python.target_types import ConsoleScript
+from pants.core.goals.generate_lockfiles import GenerateToolLockfileSentinel
+from pants.core.util_rules.config_files import ConfigFilesRequest
+from pants.engine.rules import Rule, collect_rules, rule
+from pants.engine.unions import UnionRule
+from pants.option.option_types import ArgsListOption, SkipOption
+from pants.util.docutil import git_url
+from pants.util.strutil import softwrap
+
+
+class ClangFormat(PythonToolBase):
+    options_scope = "clang-format"
+    name = "ClangFormat"
+    help = softwrap(
+        """
+        The clang-format utility for formatting C/C++ (and others) code
+        (https://clang.llvm.org/docs/ClangFormat.html).
+        """
+    )
+
+    default_version = "clang-format==14.0.3"
+    default_main = ConsoleScript("clang-format")
+
+    register_interpreter_constraints = True
+    default_interpreter_constraints = ["CPython>=3.7,<4"]
+
+    skip = SkipOption("fmt", "lint")
+    args = ArgsListOption(example="--version")
+
+    register_lockfile = True
+    default_lockfile_resource = ("pants.backend.cc.lint.clangformat", "clangformat.lock")
+    default_lockfile_path = "src/python/pants/backend/cc/lint/clangformat/clangformat.lock"
+    default_lockfile_url = git_url(default_lockfile_path)
+
+    def config_request(self, dirs: Iterable[str]) -> ConfigFilesRequest:
+        """clang-format will use the closest configuration file to the file currently being
+        formatted, so add all of them."""
+        config_files = (
+            ".clang-format",
+            "_clang-format",
+        )
+        check_existence = [os.path.join(d, file) for file in config_files for d in ("", *dirs)]
+        return ConfigFilesRequest(
+            discovery=True,
+            check_existence=check_existence,
+        )
+
+
+class ClangFormatLockfileSentinel(GenerateToolLockfileSentinel):
```

Comment:
Should `./pants export` export a virtual environment specifially with Clang? That can be helpful for IDE integration so you can point your IDE to the tool.

https://github.com/pantsbuild/pants/blob/2ce2d46abc95d1af266068f98a9095ad5a912f24/src/python/pants/backend/python/lint/isort/subsystem.py#L113-L156

(plus the union rule etc)

Reply:
My main question is if IDEs can be configured to point to the Python-installed binary in the `bin/` folder of the venv. I imagine they could? It's just a file like any other file. CLion and friends don't need to understand venvs.

Reply:
Good question. VSCode uses my system one right now, but I'll test it out if I can make it use my venv one by default.